src/main/java/org/elasticsearch/index/store/Store.java

```diff
@@ -638,13 +648,77 @@ public void deleteFile(String name) throws IOException {
         }
 
         private void innerClose() throws IOException {
-            super.close();
+            try {
+                super.close();
+            } finally {
+                fileLengthCache.clear();
+            }
+        }
+
+        @Override
+        public IndexOutput createOutput(final String name, IOContext context) throws IOException {
+            final IndexOutput output = super.createOutput(name, context);
+            fileLengthCache.remove(name); // protection against multiple writes
+            return new IndexOutput(output.toString()) {
+                @Override
+                public void close() throws IOException {
+                    output.close();
+                    fileLengthCache.put(name, getFilePointer());
+                }
+
+                @Override
+                public long getFilePointer() {
+                    return output.getFilePointer();
+                }
+
+                @Override
+                public long getChecksum() throws IOException {
+                    return output.getChecksum();
+                }
+
+                @Override
+                public void writeByte(byte b) throws IOException {
+                    output.writeByte(b);
+                }
+
+                @Override
+                public void writeBytes(byte[] b, int offset, int length) throws IOException {
+                    output.writeBytes(b, offset, length);
+                }
+            };
         }
 
         @Override
         public String toString() {
             return "store(" + in.toString() + ")";
         }
+
+        @Override
+        public void renameFile(String source, String dest) throws IOException {
+            final long len = fileLength(source);
+            super.renameFile(source, dest);
+            fileLengthCache.put(dest, len);
+            fileLengthCache.remove(source);
```

Comment:
Should we call remove before put? (Was just trying to think about what would happen if source == dest)